Azure/sap-hana
# remote deployer

### Problem
Currently, tfstates are stored in local, which makes them vulnerable.

### Solution
Seperate each module.
This pr focus on deployer: manage its tfstate.

/bootstrap/sap_library will be used to:
1. provision storage accounts for storing sapbits and tfstates
2. Generate two scripts, deployer_tfstate_script.sh is used to migrate deployer's tfstate.  saplibrary_tfstate_script.sh is used to migrate saplibrary's tfstate
 
/run/sap_library will be used to config azurem as the backend of terraform and help to migrate the tfstates of sap library from local to remote storage account.

/run/sap_deployer will be used to config azurem as the backend of terraform and help to migrate the tfstates of deployer from local to remote storage account.

### Test
1. Logon to deployer vm, (checkout to this git branch). cd ~/sap-hana/deploy/terraform/bootstrap/sap_library
2. Edit sap_library.json.
3. Execute: terraform init
4. Execute: terraform apply -auto-approve -var-file="sap_library.json"
5. After execution, there will be a bash script "saplibrary_tfstate_script.sh" generated under directory ".../run/sap_library/default/". And there will be a bash script "deployer_tfstate_script.sh" generated under directory  ".../run/sap_deployer/default/". 
6. Go to directory ".../run/sap_library/default/" in step 5, execute: ./saplibrary_tfstate_script.sh. Then, check if corresponding container (i.e. saplibrary) has the saplibrary.terraform.tfstate file.
7. Go to directory ".../run/sap_deployer/default/" in step 5, execute: ./deployer_tfstate_script.sh. Then, check if corresponding container (i.e. deployer) has the deployer.terraform.tfstate file.
